### PR TITLE
Support render targets for 'usdrender' product type

### PR DESCRIPTION
**Pull request by MustafaJafar**
## Changelog Description
Support render targets for 'usdrender' product type
This also includes
- Submitted to Deadline Husk Standalone renders (with split export jobs)
- Mark render as reviewable. 

## Additional info
It works fine on my side. 
But, for some reason I don't know yet, the jobs are submitted twice ?! 
image

## Testing notes:
1. Publish `usdrender` from Houdini using the four options:
  - Local machine rendering
  - Use existing frames (local)
  - Farm Rendering
  - Farm Rendering - Split export & render jobs
2. Also test `Review` _mark as reviewable._

### CODE REVIEW

File: server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py [hunk 1]
```diff
@@ -26,6 +26,7 @@ class ValidateUSDRenderProductPaths(pyblish.api.InstancePlugin):
     hosts = ["houdini"]
     label = "Validate USD Render Product Paths"
     optional = True
+    enabled = False
```
Comment (BigRoy) [inline server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py:29]: Does this just hint that this should be configurable in studio settings and disabled by default? Why are you disabling it?

File: server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_arnold.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def process(self, instance):
         )
         render_chunk_size = submission_data.get("chunk", 1)
         export_chunk_size = submission_data.get("export_chunk", 1)
-        usd_file_per_frame = "$F" in instance.data["ifdFile"]
+        usd_file_per_frame = instance.data["farm"]
```
Comment (BigRoy) [inline server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_arnold.py:38]: This change is incorrect, sorry. We render our USD render `usd` files as singular files, one big chunk instead of per frame - even on the farm because the files are smaller relatively (no duplication of the same data structure) and the server has better "in-memory" load of the one file and the USD files from render scene is usually small due to containing 99% references/sublayers instead of local geometry.

File: server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py [hunk 3]
```diff
@@ -131,12 +132,16 @@ def replace(match):
             self.log.debug("Render Product %s%s", aov_label, prim_path)
             self.log.debug("Product name: %s", filename)
 
+
+        # As long as we have one AOV then multipartExr should be True.
+        instance.data["multipartExr"] = len(files_by_product) <= 1
```
Comment (BigRoy) [inline server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py:137]: This is technically not entirely true. It may just not have multiple AOVs and is then not multipart, right? USD Render Products are also completely capable of writing no beauty but only e.g. a Z-pass (even though you'd usually not do that I suppose)
Out of curiosity, why do we need this?

File: server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py [hunk 4]
```diff
@@ -25,7 +25,7 @@ class CollectRenderProducts(plugin.HoudiniInstancePlugin):
     """
 
     label = "Collect Render Products"
-    order = pyblish.api.CollectorOrder + 0.4
+    order = pyblish.api.CollectorOrder + 0.04
```
Comment (BigRoy) [inline server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py:28]: Why is this needed? (I also wonder why it was +0.4 originally now)

File: client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py [hunk 5]
```diff
@@ -289,7 +292,7 @@ def get_plugin_info(self, job_type=None):
         instance = self._instance
         context = instance.context
 
-        hou_major_minor = hou.applicationVersionString().rsplit(".", 1)[0]
+        self.hou_major_minor = hou.applicationVersionString().rsplit(".", 1)[0]
```
Comment (BigRoy) [inline client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:295]: I really don't like us putting this on `self` here. Can we just pass `version` to the `_get_husk_standalone_plugin_info` function explicitly?

File: server_addon/houdini/client/ayon_houdini/plugins/publish/extract_render.py [hunk 6]
```diff
@@ -32,6 +33,8 @@ def process(self, instance):
                 rop_node.setParms({"RS_archive_enable": 1})
             elif product_type == "vray_rop":
                 rop_node.setParms({"render_export_mode": "2"})
+            elif product_type == "usdrender":
+                rop_node.setParms({"runcommand": 0})
```
Comment (BigRoy) [inline server_addon/houdini/client/ayon_houdini/plugins/publish/extract_render.py:37]: Seeing these parms being changes in the extractor and stay that way really scares me. We should usually not at all change the user's workfile - I understand the reasoning, but it goes against all of the concepts of Pyblish and thus also how publishing in OpenPype and then AYON has been structured.

We should really track this as an issue, in this case I'm fine with the logic if it works because it aligns with existing code that's there - but we should really find better ways to do this then 'breaking' the user's scene.

File: client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py [hunk 7]
```diff
@@ -71,6 +71,7 @@ class HuskStandalonePluginInfo():
     PostFrame = attr.ib(default="")
     PostRender = attr.ib(default="")
     RestartDelegate = attr.ib(default="")
+    Version = attr.ib(default="")
```
Comment (BigRoy) [inline client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:74]: I'm fine with this change here - just wanted to remark that this may have been better as a separate PR since it's a completely isolated change to also submit along with the version.

**Review by BigRoy — COMMENTED** [review]
Nice stuff - some small comments.

File: client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py [hunk 8]
```diff
@@ -96,7 +97,9 @@ class HoudiniSubmitDeadline(
                 "arnold_rop",
                 "mantra_rop",
                 "karma_rop",
-                "vray_rop"]
+                "vray_rop",
+                "usdrender"]
```
Comment (BigRoy) [inline client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:101]: How come this is new in this PR by the way? Wasn't submission to deadline already working? :/

Comment (BigRoy) [reply client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:101]: I've added it upstream in my branch as well - seemed like a refactor issue that it got lost?

File: client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py (hunk 7 above)
Comment (BigRoy) [reply client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:74]: Implemented this upstream currently. Thanks!

File: client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py (hunk 5 above)
Comment (BigRoy) [reply client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:295]: Implemented this upstream currently. Thanks!

File: client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py (hunk 8 above)
Comment (BigRoy) [reply client/ayon_core/modules/deadline/plugins/publish/submit_houdini_render_deadline.py:101]: Ah, nevermind. Found it - it's because it has its own class at the bottom of this file.
This family added now is also why you're getting double submissions @MustafaJafar

File: server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py (hunk 1 above)
Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py:29]: > Does this just hint that this should be configurable in studio settings and disabled by default?

I think yes. tbh, I have no idea what is this `$HIP/renders/$HIPNAME:r/$OS/$HIPNAME:r.$OS.$F4.exr` and it was not in the node on creation by default. and that note encouraged me to just disable the validator 😅
```
(colorbleed): This workflow might be relatively Colorbleed-specific
```

File: server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_arnold.py (hunk 2 above)
Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_arnold.py:38]: Oh, I miss interpreted the meaning of that line! 
So, in the current implementation `"ifdFile"` exists only if the render target was set to `"farm_split"` aka `"Farm Rendering - Split export & render jobs"`. 
So, this validator will crash for other render targets. because of a key error.

File: server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py (hunk 3 above)
Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py:137]: > why do we need this?

Well, searching for `"multiPartExr"` in this file [pyblish_functions.py](https://github.com/ynput/ayon-core/blob/develop/client/ayon_core/pipeline/farm/pyblish_functions.py) can answer the question. 

As far as I was able to understand the logic. if there's one AOV, the logic will assume it's reviewable.
And I was merely trying to adopt the same logic.

File: server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py (hunk 4 above)
Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py:28]: > why it was +0.4 originally

I don't know and  I didn't go through this plugin before as it's used for `usdrender` product type which didn't work on my side (before your PR). 

> Why is this needed?

because of `collect local render instances` expects some data the are collected by `collect render products`. 

https://github.com/ynput/ayon-core/blob/a8af05275b4d503847a8b9241018343f4a938b3e/server_addon/houdini/client/ayon_houdini/plugins/publish/collect_local_render_instances.py#L18-L19

File: server_addon/houdini/client/ayon_houdini/plugins/publish/extract_render.py (hunk 6 above)
Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/extract_render.py:37]: I don't like it too.  
Yes, it aligns with existing code that's there... but, I wrote that existing code and I don't like that existing code either. 
That's why I created this issue https://github.com/ynput/ayon-core/issues/417 and made suggested a PR https://github.com/ynput/ayon-core/pull/418 which made the creators more complex.

Comment (BigRoy) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/extract_render.py:37]: Ah - issue already exists, ignore me!

File: server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py (hunk 4 above)
Comment (BigRoy) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py:28]: If it works with this collector order for the different submission types - fine with me.

Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/collect_render_products.py:28]: I didn't notice any difference with different render targets.

File: server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py (hunk 1 above)
Comment (BigRoy) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py:29]: Yup, so basically what happens is that the USD Render ROP writes out an intermediate `__render__.usd` file or alike - however, by default that's just a static path, not a render file for your particular workfile if I recall correctly. And thus, if the user quickly submits to deadline three files from the same work folder it basically means that the each job would overwrite that `render.usd` file the other two jobs are also writing - hence you can't submit multiples at once. By making sure those files have unique names (e.g. containing `$HIPNAME` that solves that particular issue).

Comment (MustafaJafar) [reply server_addon/houdini/client/ayon_houdini/plugins/publish/validate_usd_render_product_paths.py:29]: I think I get it now. it important when publishing multiple `usdrender` instances at once. 

image
image

**Comment by MustafaJafar** [comment]
Closing in favor of https://github.com/BigRoy/ayon-core/pull/10
